osbuild/util/pathfd.py

```diff
@@ -0,0 +1,716 @@
+                        # unlinked file.
+                        continue
+                    else:
+                        raise e
+        finally:
+            if entries is not None:
+                entries.close()
+            _os.close(dir_fd)
+
+    def traverse(self, *, postorder = False, fn_enumerate = enumerate):
+        """Traverse Directory Tree
+
+        This creates a generator that iteratively traverses a directory tree,
+        yielding an object for every entry in that file-system tree. It
+        operates through `enumerate()` internally, but traveses an entire tree
+        rather than just a single directory level.
+
+        Note that `follow_symlink` is not provided as an option, and this
+        function currently behaves as if it was set to `False`. Traversing a
+        tree and following symlinks easily deadlocks. Furthermore, the
+        semantics are not entirely clear. Therefore, it is left to the caller
+        to resolve symlinks, if they so desire.
+
+        No entry for the object this is called on is yielded. That is, if the
+        directory that `self` points to is empty, this will yield no entries.
+
+        Parameters
+        ----------
+        postorder : Boolean, optional
+            Whether to yield entries in postorder. Default behavior is
+            preorder. That is, in preorder mode a directory is yielded before
+            the entries of the directory are yielded. In postorder mode, this
+            behavior is reversed.
+        fn_enumerate : Function, optional
+            Override the function used to enumerate a directory. By default,
+            this uses the `enumerate()` method on path-descriptors.
+
+        Returns
+        -------
+        Generator
+            A generator that produces tuples of a path-descriptor and an
+            `os.DirEntry` object is returned. Note that the directory entry
+            only contains information for the level it is on. No information
+            about the parent directories is included.
+
+        Raises
+        ------
+        OSError
+            This uses `enumerate()` internally, and thus returns the same set
+            of errors. No additional considerations apply.
+        """
+
+        assert self.is_open()
+
+        # We use a non-recursive implementation that stores every directory we
+        # recurse into on the stack, and pops it once fully enumerated.
+        stack = [ (self, None, fn_enumerate(self)) ]
+
+        while len(stack) > 0:
+            (current_dirfd, current_entry, current_scan) = stack[-1]
+
+            for (sub_dirfd, sub_entry) in current_scan:
+                if sub_dirfd.is_directory():
+                    if not postorder:
+                        yield (sub_dirfd, sub_entry)
+
+                    stack.append((sub_dirfd, sub_entry, fn_enumerate(sub_dirfd)))
+                    break
+                else:
+                    yield (sub_dirfd, sub_entry)
+            else:
```

Comment:
Probably the first time I see `for ... else` actually used.

Reply:
I read it in the python standard-library and it is really handy sometimes.